Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,700 @@ void LoadSettings()
+{
+	const auto& dockMode = mobiledockClients[carrierID].dockMode;
+	if (dockMode == ALLOW_ALL)
+	{
+		StartDockingProcedure(dockingID, carrierID);
+	}
+	else if (dockMode == ALLOW_NONE)
+	{
+		mapPendingDockingRequests[dockingID] = carrierID;
+		PrintUserCmdText(carrierID, L"%s is requesting to dock, authorise with /allowdock", Players.GetActiveCharacterName(dockingID));
+		PrintUserCmdText(dockingID, L"Docking request sent to %s", Players.GetActiveCharacterName(carrierID));
+	}
+	else if (dockMode == ALLOW_GROUP)
+	{
+		list<GROUP_MEMBER> lstGrpMembers;
+		HkGetGroupMembers((const wchar_t*)Players.GetActiveCharacterName(dockingID), lstGrpMembers);
 
-			// Remove charname from carrier's mapDockedShips.
-			mobiledockClients[carrier_client].mapDockedShips.erase(clientName);
-			mobiledockClients[carrier_client].iDockingModulesAvailable++;
+		bool isGroupMember = false;
+		for (auto& member : lstGrpMembers)
+		{
+			if (member.iClientID == carrierID)
+			{
+				isGroupMember = true;
+				break;
+			}
+		}
+		if (isGroupMember)
+		{
+			StartDockingProcedure(dockingID, carrierID);
 		}
-		// If not - redirect to proxy base in carrier's system.
 		else
 		{
-			mobiledockClients[client].carrierSystem = -1;
+			mapPendingDockingRequests[dockingID] = carrierID;
+			PrintUserCmdText(carrierID, L"%s is requesting to dock, authorise with /allowdock", Players.GetActiveCharacterName(dockingID));
+			PrintUserCmdText(dockingID, L"Docking request sent to %s", Players.GetActiveCharacterName(carrierID));
 		}
 	}
 }
 
-bool __stdcall LaunchPosHook(uint space_obj, struct CEqObj &p1, Vector &pos, Matrix &rot, int dock_mode)
+void BeamToCarrier(uint dockedID, uint carrierShipID)
 {
-	returncode = DEFAULT_RETURNCODE;
+	//teleport the player using same method as jumpdrives
+	CUSTOM_JUMP_CALLOUT_STRUCT jumpData;
+	Vector pos;
+	Matrix ori;
+	uint iSystemID;
 
-	// Redirect the ship to carrier's position. Can bug with POB plugin, changes may be required.
-	if (undockingShip.proxyBaseID == space_obj)
-	{
-		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+	pub::SpaceObj::GetLocation(carrierShipID, pos, ori);
+	pub::SpaceObj::GetSystem(carrierShipID, iSystemID);
+
+	pos.x += ori.data[0][1] * set_iMobileDockOffset;
+	pos.y += ori.data[1][1] * set_iMobileDockOffset;
+	pos.z += ori.data[2][1] * set_iMobileDockOffset;
+
+	jumpData.iClientID = dockedID;
+	jumpData.iSystemID = iSystemID;
+	jumpData.pos = pos;
+	jumpData.ori = ori;
 
-		rot = undockingShip.carrierRot;
-		pos = undockingShip.carrierPos;
-		undockingShip.proxyBaseID = 0;
+	Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_JUMP_CALLOUT, &jumpData);
+}
+
+bool CanDockOnCarrier(uint dockingID, uint carrierShipID)
+{
+	if (mobiledockClients[carrierShipID].iDockingModulesAvailable > 0)
+	{
+		return true;
 	}
-	return true;
+	if (!idToCarrierInfoMap.count(carrierShipID))
+	{
+		return false;
+	}
+	wstring dockingName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(dockingID));

```

Comment:
So the way this plugin works is that once the fighter player docks on the carrier, it's considered their 'respawn point', and they take up the spot until they're forcibly jettisoned by the carrier, or the fighter docks on another station.
As such, they must be capable of docking on a carrier if they're already 'registered' to it, since they have a reserved spot.